File: scripts/test_all_endpoints.py
#!/usr/bin/env python3
"""
Test All Bridge Service API Endpoints
"""

import asyncio
import httpx
import json
from datetime import datetime

API_BASE = "http://localhost:8000"
BRIDGE_SERVICE_BASE = "http://localhost:8000/api/bridge-service"

# Test token (would normally come from auth)
TEST_TOKEN = None  # Set if you have a test token


async def test_endpoint(name, method, url, data=None, headers=None, expected_status=None):
    """Test a single endpoint"""
    print(f"\n{'='*70}")
    print(f"TEST: {name}")
    print(f"{'='*70}")
    print(f"Method: {method}")
    print(f"URL: {url}")
    
    if data:
        print(f"Body: {json.dumps(data, indent=2)[:200]}...")
    
    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            if method == "GET":
                response = await client.get(url, headers=headers)
            elif method == "POST":
                response = await client.post(url, json=data, headers=headers)
            else:
                response = await client.request(method, url, json=data, headers=headers)
            
            print(f"Status: {response.status_code}")
            
            try:
                response_data = response.json()
                print(f"Response: {json.dumps(response_data, indent=2)[:500]}...")
            except:
                print(f"Response: {response.text[:500]}...")
            
            if expected_status:
                if response.status_code == expected_status:
                    print(f"[PASS] Expected status {expected_status}")
                    return True
                else:
                    print(f"[WARN] Status {response.status_code} (expected {expected_status})")
                    return response.status_code in [200, 201, 401, 403]  # Accept auth errors
            else:
                if response.status_code in [200, 201]:
                    print(f"[SUCCESS]")
                    return True
                elif response.status_code in [401, 403]:
                    print(f"[AUTH] Authentication required (expected)")
                    return True  # Count as success since endpoint exists
                else:
                    print(f"[FAIL] Status {response.status_code}")
                    return False
                    
    except httpx.ConnectError:
        print(f"[ERROR] Cannot connect to {url}")
        print(f"   Make sure the backend is running on port 8000")
        return False
    except Exception as e:
        print(f"❌ ERROR: {e}")
        return False


async def test_all_endpoints():
    """Test all bridge service endpoints"""
    print("\n" + "=" * 70)
    print("BRIDGE SERVICE API ENDPOINTS - COMPREHENSIVE TEST")
    print("=" * 70)
    print(f"\nTesting API at: {API_BASE}")
    print(f"Time: {datetime.now().isoformat()}")
    
    headers = {}
    if TEST_TOKEN:
        headers["Authorization"] = f"Bearer {TEST_TOKEN}"
    
    results = []
    
    # Health and Status Endpoints
    print("\n" + "="*70)
    print("HEALTH & STATUS ENDPOINTS")
    print("="*70)
    
    results.append(("Health Check", await test_endpoint(
        "Bridge Service Health",
        "GET",
        f"{BRIDGE_SERVICE_BASE}/health",
        expected_status=200
    )))
    
    results.append(("Paradox API Health", await test_endpoint(
        "Paradox API Health",
        "GET",
        f"{API_BASE}/health",
        expected_status=200
    )))
    
    results.append(("Paradox API Root", await test_endpoint(
        "Paradox API Root",
        "GET",
        f"{API_BASE}/",
        expected_status=200
    )))
    
    # Network Endpoints
    print("\n" + "="*70)
    print("NETWORK ENDPOINTS")
    print("="*70)
    
    results.append(("Get Supported Networks", await test_endpoint(
        "Get Supported Networks",
        "GET",
        f"{BRIDGE_SERVICE_BASE}/network/supported",
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    results.append(("Get Network Status (All)", await test_endpoint(
        "Get Network Status (All)",
        "GET",
        f"{BRIDGE_SERVICE_BASE}/network/status",
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    results.append(("Get Network Status (Ethereum)", await test_endpoint(
        "Get Network Status (Ethereum)",
        "GET",
        f"{BRIDGE_SERVICE_BASE}/network/status?network=ethereum",
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    # Bridge Analysis Endpoints
    print("\n" + "="*70)
    print("BRIDGE ANALYSIS ENDPOINTS")
    print("="*70)
    
    test_bridge_address = "0x1234567890123456789012345678901234567890"
    
    results.append(("List Bridges", await test_endpoint(
        "List Bridges",
        "GET",
        f"{BRIDGE_SERVICE_BASE}/list?limit=5",
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    results.append(("Get Bridge Info", await test_endpoint(
        "Get Bridge Info",
        "GET",
        f"{BRIDGE_SERVICE_BASE}/{test_bridge_address}/info",
        headers=headers,
        expected_status=[200, 401, 403, 404]
    )))
    
    results.append(("Analyze Bridge", await test_endpoint(
        "Analyze Bridge",
        "POST",
        f"{BRIDGE_SERVICE_BASE}/analyze",
        data={
            "bridge_address": test_bridge_address,
            "source_network": "ethereum",
            "target_network": "polygon",
            "analysis_depth": "comprehensive"
        },
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    results.append(("Get Security Score", await test_endpoint(
        "Get Security Score",
        "POST",
        f"{BRIDGE_SERVICE_BASE}/security-score",
        data={
            "bridge_address": test_bridge_address,
            "network": "ethereum"
        },
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    results.append(("Get Bridge Metrics", await test_endpoint(
        "Get Bridge Metrics",
        "GET",
        f"{BRIDGE_SERVICE_BASE}/metrics?time_range=7d",
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    # Advanced Security Endpoints
    print("\n" + "="*70)
    print("ADVANCED SECURITY ENDPOINTS")
    print("="*70)
    
    results.append(("Detect Attestation Anomalies", await test_endpoint(
        "Detect Attestation Anomalies",
        "POST",
        f"{BRIDGE_SERVICE_BASE}/detect-attestation-anomalies",
        data={
            "bridge_address": test_bridge_address,
            "network": "ethereum",
            "time_range": "24h",
            "include_details": True
        },
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    results.append(("Analyze Quorum Skews", await test_endpoint(
        "Analyze Quorum Skews",
        "POST",
        f"{BRIDGE_SERVICE_BASE}/analyze-quorum-skews",
        data={
            "bridge_address": test_bridge_address,
            "network": "ethereum",
            "analysis_period": "7d"
        },
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    results.append(("Proof of Reserves", await test_endpoint(
        "Proof of Reserves",
        "POST",
        f"{BRIDGE_SERVICE_BASE}/proof-of-reserves",
        data={
            "bridge_address": test_bridge_address,
            "network": "ethereum",
            "include_asset_breakdown": True
        },
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    results.append(("Comprehensive Security Scan", await test_endpoint(
        "Comprehensive Security Scan",
        "POST",
        f"{BRIDGE_SERVICE_BASE}/comprehensive-scan",
        data={
            "bridge_address": test_bridge_address,
            "network": "ethereum",
            "transaction_data": [],
            "scan_options": {
                "include_attack_analysis": True,
                "include_signature_analysis": True,
                "deep_scan": False
            }
        },
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    # Transaction Endpoints
    print("\n" + "="*70)
    print("TRANSACTION ENDPOINTS")
    print("="*70)
    
    test_tx_hash = "0xabcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890"
    
    results.append(("Validate Transaction", await test_endpoint(
        "Validate Transaction",
        "POST",
        f"{BRIDGE_SERVICE_BASE}/transaction/validate",
        data={
            "transaction_hash": test_tx_hash,
            "network": "ethereum"
        },
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    results.append(("Get Transaction Status", await test_endpoint(
        "Get Transaction Status",
        "GET",
        f"{BRIDGE_SERVICE_BASE}/transaction/{test_tx_hash}/status",
        headers=headers,
        expected_status=[200, 401, 403, 404]
    )))
    
    # Vulnerability Endpoints
    print("\n" + "="*70)
    print("VULNERABILITY ENDPOINTS")
    print("="*70)
    
    results.append(("Scan Vulnerabilities", await test_endpoint(
        "Scan Vulnerabilities",
        "POST",
        f"{BRIDGE_SERVICE_BASE}/vulnerability/scan",
        data={
            "contract_addresses": [test_bridge_address],
            "networks": ["ethereum", "polygon"],
            "scan_type": "comprehensive"
        },
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    # Security Monitoring Endpoints
    print("\n" + "="*70)
    print("SECURITY MONITORING ENDPOINTS")
    print("="*70)
    
    results.append(("Get Security Dashboard", await test_endpoint(
        "Get Security Dashboard",
        "GET",
        f"{BRIDGE_SERVICE_BASE}/security/dashboard",
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    results.append(("Get Security Events", await test_endpoint(
        "Get Security Events",
        "GET",
        f"{BRIDGE_SERVICE_BASE}/security/events?limit=10",
        headers=headers,
        expected_status=[200, 401, 403]
    )))
    
    # Summary
    print("\n" + "="*70)
    print("TEST SUMMARY")
    print("="*70)
    
    passed = sum(1 for _, result in results if result)
    total = len(results)
    
    print(f"\nTotal Endpoints Tested: {total}")
    print(f"Successful: {passed}")
    print(f"Failed: {total - passed}")
    print(f"Success Rate: {(passed/total)*100:.1f}%")
    
    print("\nDetailed Results:")
    for name, result in results:
        status = "[PASS]" if result else "[FAIL]"
        print(f"  {status}: {name}")
    
    print("\n" + "="*70)
    
    if passed == total:
        print("\n[SUCCESS] ALL ENDPOINTS ACCESSIBLE!")
        print("   All endpoints are responding correctly.")
    elif passed >= total * 0.8:
        print(f"\n[OK] MOST ENDPOINTS WORKING ({passed}/{total})")
        print("   Some endpoints may require authentication.")
        print("   Set TEST_TOKEN environment variable to test authenticated endpoints.")
    else:
        print(f"\n[WARN] SOME ENDPOINTS FAILING ({passed}/{total})")
        print("   Check that:")
        print("   1. Backend is running on port 8000")
        print("   2. Bridge service is accessible")
        print("   3. Endpoints are properly registered")
    
    print("="*70)
    
    return passed, total


if __name__ == "__main__":
    print("\nStarting comprehensive endpoint tests...")
    print("Make sure the backend is running on http://localhost:8000\n")
    
    passed, total = asyncio.run(test_all_endpoints())
    
    exit(0 if passed >= total * 0.8 else 1)

File: scripts/test-all-endpoints.ts
#!/usr/bin/env tsx
/**
 * Comprehensive Endpoint Test Suite
 * Tests ALL API endpoints from every service
 * Continues until all endpoints return 200
 */

// Use native fetch (Node 18+) or node-fetch if available
const fetchFn = typeof fetch !== 'undefined' ? fetch : require('node-fetch');

const API_BASE = process.env.API_BASE_URL || 'http://localhost:8000';
const API_KEY = process.env.API_KEY || 'demo-api-key';
// Get AUTH_TOKEN from env or command line args
const AUTH_TOKEN = process.env.AUTH_TOKEN || process.argv.find(arg => arg.startsWith('--token='))?.split('=')[1] || '';

interface Endpoint {
  method: string;
  path: string;
  name: string;
  requiresAuth?: boolean;
  body?: any;
  params?: Record<string, string>;
}

const endpoints: Endpoint[] = [
  // Root & Health
  { method: 'GET', path: '/', name: 'Root endpoint' },
  { method: 'GET', path: '/health', name: 'Health check' },
  
  // Account Management
  { method: 'GET', path: '/api/account/', name: 'Get account info', requiresAuth: true },
  { method: 'GET', path: '/api/account/wallets', name: 'Get user wallets', requiresAuth: true },
  { method: 'POST', path: '/api/account/wallets', name: 'Create wallet', requiresAuth: true, body: { name: 'Test Wallet', network: 'ethereum' } },
  { method: 'GET', path: '/api/account/devices', name: 'Get devices', requiresAuth: true },
  { method: 'GET', path: '/api/account/sessions', name: 'Get sessions', requiresAuth: true },
  
  // Wallet Operations
  { method: 'GET', path: '/api/wallet/status', name: 'Get wallet status', requiresAuth: true },
  
  // Bridge Operations
  { method: 'GET', path: '/api/bridge/networks', name: 'Get bridge networks', requiresAuth: true },
  { method: 'GET', path: '/api/bridge/fee', name: 'Get bridge fee', requiresAuth: true },
  
  // MEV Protection
  { method: 'GET', path: '/api/mev/status', name: 'Get MEV status', requiresAuth: true },
  { method: 'GET', path: '/api/mev/stats', name: 'Get MEV stats', requiresAuth: true },
  
  // Vault Management
  { method: 'GET', path: '/api/vault/user/vaults', name: 'Get user vaults', requiresAuth: true },
  { method: 'GET', path: '/api/vault/stats/global', name: 'Get global vault stats' },
  
  // Security
  { method: 'GET', path: '/api/security/mpc', name: 'Get MPC metrics', requiresAuth: true },
  { method: 'GET', path: '/api/security/events', name: 'Get security events', requiresAuth: true },
  { method: 'GET', path: '/api/security/wallet-guard/threats', name: 'Get Wallet Guard threats', requiresAuth: true },
  
  // AI Assistant
  { method: 'GET', path: '/api/ai/health', name: 'AI health check' },
  { method: 'POST', path: '/api/ai/chat', name: 'AI chat', requiresAuth: true, body: { message: 'Hello', conversationHistory: [] } },
  
  // Transaction History
  { method: 'GET', path: '/api/transactions/', name: 'Get transactions', requiresAuth: true },
  { method: 'GET', path: '/api/transactions/stats/summary', name: 'Get transaction stats', requiresAuth: true },
  
  // Settings
  { method: 'GET', path: '/api/settings/', name: 'Get settings', requiresAuth: true },
  { method: 'GET', path: '/api/settings/app-version', name: 'Get app version' },
  
  // Notifications
  { method: 'GET', path: '/api/notifications/', name: 'Get notifications', requiresAuth: true },
  { method: 'GET', path: '/api/notifications/badge-count', name: 'Get badge count', requiresAuth: true },
  
  // Biometric
  { method: 'GET', path: '/api/biometric/status', name: 'Get biometric status', requiresAuth: true },
  
  // Support
  { method: 'GET', path: '/api/support/help', name: 'Get help articles' },
  { method: 'GET', path: '/api/support/status', name: 'Get system status' },
  
  // Legal
  { method: 'GET', path: '/api/legal/terms-of-service', name: 'Get terms of service' },
  { method: 'GET', path: '/api/legal/privacy-policy', name: 'Get privacy policy' },
  { method: 'GET', path: '/api/legal/versions', name: 'Get legal versions' },
  
  // Fiat On-Ramp
  { method: 'GET', path: '/api/fiat/providers', name: 'Get fiat providers' },
  
  // DEX Aggregator
  { method: 'POST', path: '/api/swaps/aggregators', name: 'Get swap aggregator quotes', requiresAuth: true, body: { fromToken: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE', toToken: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48', amount: '1000000000000000000', fromAddress: '0x0000000000000000000000000000000000000000', chainId: 1 } },
  
  // Cross-Chain Swaps
  { method: 'POST', path: '/api/cross-chain/routes', name: 'Get cross-chain routes', requiresAuth: true, body: { fromChainId: 1, toChainId: 137, fromToken: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE', toToken: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE', amount: '1000000000000000000', fromAddress: '0x0000000000000000000000000000000000000000' } },
  
  // NFT Gallery
  { method: 'POST', path: '/api/nft/gallery', name: 'Get NFT gallery', requiresAuth: true, body: { walletAddress: '0x0000000000000000000000000000000000000000' } },
  
  // Market Data
  { method: 'POST', path: '/api/market/prices', name: 'Get market prices', requiresAuth: true, body: { tokens: [{ address: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE', chainId: 1 }] } },
  { method: 'GET', path: '/api/market/pendle/markets', name: 'Get Pendle markets', requiresAuth: true },
  
  // MEV Protection Service (v1)
  { method: 'GET', path: '/api/v1/protection/status', name: 'Get protection status', requiresAuth: true },
  { method: 'GET', path: '/api/v1/stats', name: 'Get protection stats', requiresAuth: true },
  { method: 'GET', path: '/api/v1/dashboard', name: 'Get dashboard', requiresAuth: true },
  { method: 'GET', path: '/api/v1/networks', name: 'Get networks', requiresAuth: true },
  { method: 'GET', path: '/api/v1/threats', name: 'Get threats', requiresAuth: true },
  { method: 'GET', path: '/api/v1/relays', name: 'Get relays', requiresAuth: true },
  { method: 'GET', path: '/api/v1/mev/metrics', name: 'Get MEV metrics', requiresAuth: true },
  
  // Enhanced MEV API
  { method: 'GET', path: '/status', name: 'Get protection status (enhanced)' },
  { method: 'GET', path: '/api/v1/kpi/metrics', name: 'Get KPI metrics', requiresAuth: true },
  { method: 'GET', path: '/api/v1/analytics/dashboard', name: 'Get analytics dashboard', requiresAuth: true },
  
  // Cross-Chain Bridge Service
  { method: 'GET', path: '/api/v1/bridge/metrics', name: 'Get bridge metrics' },
  { method: 'GET', path: '/api/v1/bridge/list', name: 'List bridges' },
  { method: 'GET', path: '/api/v1/network/status', name: 'Get network status' },
  { method: 'GET', path: '/api/v1/network/supported', name: 'Get supported networks' },
  { method: 'GET', path: '/api/v1/vulnerability/threats', name: 'Get vulnerability threats' },
  { method: 'GET', path: '/api/v1/security/dashboard', name: 'Get security dashboard', requiresAuth: true },
];

interface TestResult {
  endpoint: Endpoint;
  status: number;
  success: boolean;
  error?: string;
  responseTime: number;
}

async function testEndpoint(endpoint: Endpoint): Promise<TestResult> {
  const startTime = Date.now();
  const url = new URL(`${API_BASE}${endpoint.path}`);
  
  // Add query params if provided
  if (endpoint.params) {
    Object.entries(endpoint.params).forEach(([key, value]) => {
      url.searchParams.append(key, value);
    });
  }
  
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    'X-API-Key': API_KEY,
  };
  
  if (endpoint.requiresAuth && AUTH_TOKEN) {
    headers['Authorization'] = `Bearer ${AUTH_TOKEN}`;
  }
  
  try {
    const response = await fetch(url.toString(), {
      method: endpoint.method,
      headers,
      body: endpoint.body ? JSON.stringify(endpoint.body) : undefined,
    });
    
    const responseTime = Date.now() - startTime;
    const success = response.status === 200;
    
    let error: string | undefined;
    if (!success) {
      try {
        const errorData = await response.json();
        error = errorData.error || errorData.detail || `Status ${response.status}`;
      } catch {
        error = `Status ${response.status}`;
      }
    }
    
    return {
      endpoint,
      status: response.status,
      success,
      error,
      responseTime,
    };
  } catch (err) {
    const responseTime = Date.now() - startTime;
    return {
      endpoint,
      status: 0,
      success: false,
      error: err instanceof Error ? err.message : 'Unknown error',
      responseTime,
    };
  }
}

async function runTests() {
  console.log('🚀 Starting comprehensive endpoint tests...\n');
  console.log(`API Base URL: ${API_BASE}`);
  console.log(`Total endpoints to test: ${endpoints.length}\n`);
  
  const results: TestResult[] = [];
  let passed = 0;
  let failed = 0;
  let skipped = 0;
  
  for (const endpoint of endpoints) {
    // Skip auth-required endpoints if no token provided
    if (endpoint.requiresAuth && !AUTH_TOKEN) {
      console.log(`⏭️  SKIPPED: ${endpoint.name} (${endpoint.method} ${endpoint.path}) - No auth token`);
      skipped++;
      continue;
    }
    
    process.stdout.write(`Testing: ${endpoint.name}... `);
    const result = await testEndpoint(endpoint);
    results.push(result);
    
    if (result.success) {
      console.log(`✅ PASSED (${result.status}) - ${result.responseTime}ms`);
      passed++;
    } else {
      console.log(`❌ FAILED (${result.status}) - ${result.error || 'Unknown error'} - ${result.responseTime}ms`);
      failed++;
    }
    
    // Small delay to avoid overwhelming the server
    await new Promise(resolve => setTimeout(resolve, 100));
  }
  
  console.log('\n' + '='.repeat(80));
  console.log('📊 TEST SUMMARY');
  console.log('='.repeat(80));
  console.log(`✅ Passed: ${passed}`);
  console.log(`❌ Failed: ${failed}`);
  console.log(`⏭️  Skipped: ${skipped}`);
  console.log(`📈 Success Rate: ${((passed / (passed + failed)) * 100).toFixed(2)}%`);
  console.log('='.repeat(80));
  
  if (failed > 0) {
    console.log('\n❌ FAILED ENDPOINTS:');
    results
      .filter(r => !r.success)
      .forEach(r => {
        console.log(`  - ${r.endpoint.name} (${r.endpoint.method} ${r.endpoint.path})`);
        console.log(`    Status: ${r.status}, Error: ${r.error || 'Unknown'}`);
      });
  }
  
  // Return exit code based on results
  process.exit(failed > 0 ? 1 : 0);
}

// Run tests
runTests().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

